File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/12-BST/LeetCode/LeetCode-0653-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class BSTForwardIterator {
    stack<TreeNode *> S;
public:
    BSTForwardIterator(TreeNode *root){
        while (root != NULL){S.push(root); root = root->left;}
    }
    bool hasNext(){return not S.empty();}
    int next(){
        TreeNode *p = S.top();
        int ret = p->val;
        S.pop();
        if (p->right != NULL){
            p = p->right;
            while (p != NULL){S.push(p); p=p->left;}
        } //end-if 
        return ret;       
    } // end-next
};

class BSTReverseIterator {
    stack<TreeNode *> S;
public:
    BSTReverseIterator(TreeNode *root){
        while (root != NULL){S.push(root); root = root->right;}
    }
    bool hasNext(){return not S.empty();}
    int next(){
        TreeNode *p = S.top();
        int ret = p->val;
        S.pop();
        if (p->left != NULL){
            p = p->left;
            while (p != NULL){S.push(p); p=p->right;}
        } //end-if 
        return ret;       
    } // end-next
};

class Solution {
public:
    /// Here is an O(n) time algorithm that uses O(logn) extra space
    /// The idea is to create one iterator that walks in
    /// the forward direction (in ascending order of the keys)
    /// and one iterator that walks in the backward direction (in descending order of the keys)
    bool findTarget(TreeNode* root, int k) {
        BSTForwardIterator forwardIter(root);
        BSTReverseIterator reverseIter(root);

        int num1 = forwardIter.next();
        int num2 = reverseIter.next();

        while (num1 < num2){
            int sum = num1 + num2;

            if (sum == k) return true; // Found it
            if (sum < k) num1 = forwardIter.next();
            else num2 = reverseIter.next();
        } //end-while

        return false;
    } //end-findTarget
};

File: CSCI_313-323_Data_Structures_And_Algorithms/CSCI_313_Data_Structures/ExCode/12-BST/LeetCode/LeetCode-0653-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class BSTForwardIterator implements Iterator<Integer> {
    private Stack<TreeNode> S = new Stack<>();
    public BSTForwardIterator(TreeNode root){
        while (root != null){S.push(root); root = root.left;}
    }
    @Override
    public boolean hasNext(){return !S.isEmpty();}

    @Override
    public Integer next(){
        TreeNode p = S.pop();
        int ret = p.val;
        if (p.right != null){
            p = p.right;
            while (p != null){S.push(p); p=p.left;}
        } //end-if 
        return ret;       
    } // end-next
};

class BSTReverseIterator implements Iterator<Integer> {
    private Stack<TreeNode> S = new Stack<>();
    public BSTReverseIterator(TreeNode root){
        while (root != null){S.push(root); root = root.right;}
    }
    @Override
    public boolean hasNext(){return !S.isEmpty();}

    @Override
    public Integer next(){
        TreeNode p = S.pop();
        int ret = p.val;
        if (p.left != null){
            p = p.left;
            while (p != null){S.push(p); p=p.right;}
        } //end-if 
        return ret;       
    } // end-next
};

class Solution {
    /// Here is an O(n) time algorithm that uses O(logn) extra space
    /// The idea is to create one iterator that walks in
    /// the forward direction (in ascending order of the keys)
    /// and one iterator that walks in the backward direction (in descending order of the keys)
    public boolean findTarget(TreeNode root, int k) {
        BSTForwardIterator forwardIter = new BSTForwardIterator(root);
        BSTReverseIterator reverseIter = new BSTReverseIterator(root);

        int num1 = forwardIter.next();
        int num2 = reverseIter.next();

        while (num1 < num2){
            int sum = num1 + num2;

            if (sum == k) return true; // Found it
            if (sum < k) num1 = forwardIter.next();
            else num2 = reverseIter.next();
        } //end-while

        return false;
    }
}